eth/eth2.go

```diff
@@ -0,0 +1,320 @@
+		return nil, fmt.Errorf("child timestamp lower than parent's: %d >= %d", parent.Time(), params.Timestamp)
+	}
+	// this will ensure we're not going off too far in the future
+	if now := uint64(time.Now().Unix()); params.Timestamp > now+1 {
+		wait := time.Duration(params.Timestamp-now) * time.Second
+		log.Info("Producing block too far in the future", "wait", common.PrettyDuration(wait))
+		time.Sleep(wait)
+	}
+
+	pending, err := pool.Pending()
+	if err != nil {
+		return nil, err
+	}
+
+	coinbase, err := api.eth.Etherbase()
+	if err != nil {
+		return nil, err
+	}
+	num := parent.Number()
+	header := &types.Header{
+		ParentHash: parent.Hash(),
+		Number:     num.Add(num, common.Big1),
+		Coinbase:   coinbase,
+		GasLimit:   parent.GasLimit(), // Keep the gas limit constant in this prototype
+		Extra:      []byte{},
+		Time:       params.Timestamp,
+	}
+	err = api.eth.Engine().Prepare(bc, header)
+	if err != nil {
+		return nil, err
+	}
+
+	err = api.makeEnv(parent, header)
+	if err != nil {
+		return nil, err
+	}
+	signer := types.NewEIP155Signer(bc.Config().ChainID)
+	txs := types.NewTransactionsByPriceAndNonce(signer, pending)
+
+	var transactions []*types.Transaction
+
+	for {
+		if api.env.gasPool.Gas() < chainParams.TxGas {
+			log.Trace("Not enough gas for further transactions", "have", api.env.gasPool, "want", chainParams.TxGas)
+			break
+		}
+
+		tx := txs.Peek()
+		if tx == nil {
+			break
+		}
+
+		from, _ := types.Sender(signer, tx)
+		// XXX replay protection check is missing
+
+		// Execute the transaction
+		api.env.state.Prepare(tx.Hash(), common.Hash{}, api.env.tcount)
+		err := api.commitTransaction(tx, coinbase, params.RecentBeaconBlockRoots, params.RandaoMix)
+		switch err {
+		case core.ErrGasLimitReached:
+			// Pop the current out-of-gas transaction without shifting in the next from the account
+			log.Trace("Gas limit exceeded for current block", "sender", from)
+			txs.Pop()
+
+		case core.ErrNonceTooLow:
+			// New head notification data race between the transaction pool and miner, shift
+			log.Trace("Skipping transaction with low nonce", "sender", from, "nonce", tx.Nonce())
+			txs.Shift()
+
+		case core.ErrNonceTooHigh:
+			// Reorg notification data race between the transaction pool and miner, skip account =
+			log.Trace("Skipping account with high nonce", "sender", from, "nonce", tx.Nonce())
+			txs.Pop()
+
+		case nil:
+			// Everything ok, collect the logs and shift in the next transaction from the same account
+			api.env.tcount++
+			txs.Shift()
+			transactions = append(transactions, tx)
+
+		default:
+			// Strange error, discard the transaction and get the next in line (note, the
+			// nonce-too-high clause will prevent us from executing in vain).
+			log.Debug("Transaction failed, account skipped", "hash", tx.Hash(), "err", err)
+			txs.Shift()
+		}
+	}
+
+	block, err := api.eth.Engine().FinalizeAndAssemble(bc, header, api.env.state, transactions, nil /* uncles */, api.env.receipts)
```

Comment:
QQ: the block reward is applied based on the ethash rules. So in the eth2, actually there is no block reward here right? The reward will be applied in the consensus layer(they have their own accounting model).

Reply:
Yes, this is the long term view. Right after the merge, you won't be able to move the "eth2 eth" from the beacon chain into the eth1 "execution layer", so it still makes sense to offer a reward... more importantly, for the upcoming testnet: if you don't have a way to produce value, you won't be able to test anything.